File: react/src/api/types/output.ts
import { Change, Message, User } from "./data";

export enum OutputTypes {
    Error = "Error",
    Alive = "Alive",
    CurrentState = "CurrentState",
    UserJoined = "UserJoined",
    UserLeft = "UserLeft",
    Posted = "Posted",
}

export enum OutputErrors {
    UserAlreadyJoined = "UserAlreadyJoined",
}

export interface ErrorOutput {
    type: OutputTypes.Error;
    payload: OutputErrors;
}

export interface AliveOutput {
    type: OutputTypes.Alive;
}

export interface CurrentStateOutput {
    type: OutputTypes.CurrentState;
    payload: {
        myself: User;
        users: User[];
        messages: Message[];
    };
}

export interface UserJoinedOutput {
    type: OutputTypes.UserJoined;
    payload: Change;
}

export interface UserLeftOutput {
    type: OutputTypes.UserLeft;
    payload: Change;
}

export interface PostedOutput {
    type: OutputTypes.Posted;
    payload: {
        message: Message;
    };
}

export type Output =
    | ErrorOutput
    | AliveOutput
    | CurrentStateOutput
    | UserJoinedOutput
    | UserLeftOutput
    | PostedOutput;

File: core/src/modules/output.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::data::{Message, User};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", content = "payload")]
pub enum Output {
    Error(OutputErrors),
    Alive,
    CurrentState(CurrentState),
    UserJoined(UserActivityTimestamp),
    UserLeft(UserActivityTimestamp),
    Posted(Posted),
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub enum OutputErrors {
    UserAlreadyJoined,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct CurrentState {
    myself: User,
    users: Vec<User>,
    messages: Vec<Message>,
}

impl CurrentState {
    pub fn new(myself: User, users: Vec<User>, messages: Vec<Message>) -> Self {
        CurrentState { myself, users, messages }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct UserActivityTimestamp {
    user: User,
    timestamp: DateTime<Utc>,
}

impl UserActivityTimestamp {
    pub fn new(user: User) -> Self {
        UserActivityTimestamp { user, timestamp: Utc::now() }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct Posted {
    message: Message,
}

impl Posted {
    pub fn new(message: Message) -> Self {
        Posted { message }
    }
}
